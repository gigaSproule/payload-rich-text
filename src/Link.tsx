import { CSSProperties } from "react";
import type { TextData } from "./Text";
import { Text } from "./Text";
import type { Options } from "./types";

export type DefaultCustomLinkData = Omit<LinkData, "fields"> & {
  fields: {
    url: string;
    newTab: boolean;
    linkType: "custom";
  };
};

export type DefaultInternalLinkData = Omit<LinkData, "fields"> & {
  fields: {
    url: string;
    doc: {
      relationTo: string;
      value:
        | {
            id: string;
            [key: string]: unknown;
          }
        | string;
    };
    newTab: boolean;
    linkType: "internal";
  };
};

export type LinkData = {
  type: "link";
  format: string;
  indent: number;
  version: number;
  direction: "ltr" | "rtl" | null;
  id: string;
  fields: object;
  children: TextData[];
};

const isDefaultCustomLink = (link: LinkData): link is DefaultCustomLinkData => {
  return "linkType" in link.fields && link.fields.linkType === "custom";
};

export type Props = {
  data: LinkData;
  options?: Options;
  style?: CSSProperties;
};

export const Link = ({ data, options, style }: Props) => {
  const children = data.children.map((linkChild) => (
    <Text key={linkChild.text} data={linkChild} options={options} />
  ));
  return options?.link ? (
    options.link(data, children)
  ) : isDefaultCustomLink(data) ? (
    <a
      href={data.fields.url}
      target={data.fields.newTab ? "_blank" : undefined}
      rel={data.fields.newTab ? "noopener noreferrer" : undefined}
      style={style}
    >
      {children}
    </a>
  ) : undefined;
};
