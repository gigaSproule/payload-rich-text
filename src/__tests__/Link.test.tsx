import { describe, expect, it } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import type { Props } from "../Link";
import { Link } from "../Link";
import {
  createDefaultCustomLinkData,
  createDefaultInternalLinkData,
  createLinkData,
} from "./helpers/createLinkData";

describe("Link", () => {
  it("should not render anything if no children are provided", () => {
    const linkData: Props["data"] = createDefaultCustomLinkData({
      children: [],
    });
    const { container } = render(<Link data={linkData} />);
    expect(container).toMatchSnapshot();
  });

  it("should not render anything if no options are provided for custom link", () => {
    const linkData: Props["data"] = createLinkData();
    const { container } = render(<Link data={linkData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should not render anything if no options are provided for default internal link", () => {
    const linkData: Props["data"] = createDefaultInternalLinkData();
    const { container } = render(<Link data={linkData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default a if no options are provided for default custom link", () => {
    const linkData: Props["data"] = createDefaultCustomLinkData();
    const { container } = render(<Link data={linkData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render default a with target and rel if no options are provided for default custom link and new tab is defined", () => {
    const linkData: Props["data"] = createDefaultCustomLinkData({
      fields: {
        linkType: "custom",
        url: "https://www.example.com",
        newTab: true,
      },
    });
    render(<Link data={linkData} options={{}} />);
    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("target", "_blank");
    expect(link).toHaveAttribute("rel", "noopener noreferrer");
  });

  it("should render custom render if options provided with link", () => {
    const linkData: Props["data"] = createDefaultCustomLinkData();
    const { container } = render(
      <Link
        data={linkData}
        options={{
          link: (linkData, children, style) => {
            return (
              <div>
                <span>Custom A - {JSON.stringify(linkData)}</span>
                {children}
                {JSON.stringify(style)}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
