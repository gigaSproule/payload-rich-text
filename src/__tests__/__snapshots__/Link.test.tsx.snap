// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Link should not render anything if no children are provided 1`] = `
<div>
  <a
    href="https://www.example.com"
  />
</div>
`;

exports[`Link should not render anything if no options are provided for custom link 1`] = `
<div>
  <a
    href="https://www.example.com"
  >
    heading text
  </a>
</div>
`;

exports[`Link should not render anything if no options are provided for default internal link 1`] = `<div />`;

exports[`Link should render custom render if options provided with link 1`] = `
<div>
  <div>
    <span>
      Custom A - 
      {"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}],"type":"link","format":"","indent":0,"version":1,"direction":"ltr","id":"id","fields":{"url":"https://www.example.com","newTab":false,"linkType":"custom"}}
    </span>
    heading text
  </div>
</div>
`;

exports[`Link should render default a if no options are provided for default custom link 1`] = `
<div>
  <a
    href="https://www.example.com"
  >
    heading text
  </a>
</div>
`;
